### PR TITLE
improve reqwest backend

### PR DESCRIPTION
**Pull request by seanmonstar**
- use a single Client, yay connection pooling
- fix proxies (scheme was missing)
- fix reqwest partial downloads
- add tests (including partial downloads)
- add warning about RUSTUP_USE_HYPER var

### CODE REVIEW

File: src/download/tests/download-curl-resume.rs [hunk 1]
```diff
@@ -62,36 +36,48 @@ fn partially_downloaded_file_gets_resumed_from_byte_offset() {
...
+    let callback_partial = Mutex::new(false);
+    let callback_len = Mutex::new(None);
+    let received_in_callback = Mutex::new(Vec::new());
 
     download_to_path_with_backend(Backend::Curl,
                                   &from_url,
                                   &target_path,
                                   true,
                                   Some(&|msg| {
+        println!("event: {:?}", msg);
```
Comment (alexcrichton) [inline src/download/tests/download-curl-resume.rs]: Stray println?

**Comment by alexcrichton** [comment]
Looks great to me, thanks!

File: src/download/tests/download-curl-resume.rs (hunk 1 above)
Comment (seanmonstar) [reply src/download/tests/download-curl-resume.rs]: Woops, yep. Used to learn that I needed to add a `Seek` to the reqwest backend file downloader.

**Comment by seanmonstar** [comment]
I noticed in using this that it would warn about the environment variable several times a run, as it tried to download several files. So I've added a static bool to only emit the warning once...

**Comment by seanmonstar** [comment]
I had removed a `[replace]` for mio, as it was warning me that it wasn't needed. I'm not sure why, but I've added it back, to fix those tests on Android.

**Comment by seanmonstar** [comment]
I had included downgrades in the `Cargo.lock` that I didn't even notice. I've gone through and fixed every single one, besides the absolute necessary to needed in this PR.

**Comment by alexcrichton** [comment]
Tests passed, gonna merge manually.

**Pull request merged**